### ISSUE TITLE
Pileup calling fails for downsampled BAMs

### ISSUE THREAD

**Issue by fgfrost** [HKU-BAL/Clair3#293]
Hello,

I've been having an odd issue when running Clair3 version 1.0.4 on downsampled BAMs. We've been able to run Clair3 just fine on the full minimap2-aligned BAM files for ONT sequencing data. However, when we use `samtools view` (full command below) to downsample the BAMs, we get two main issues when running Clair3.

```
samtools view --subsample 0.11111 --subsample-seed  2708 -b --threads ${SLURM_JOB_CPUS_PER_NODE} --output ${BAMDIR}/${sampleName}_${current_coverage}_${seed}.bam ${BAMDIR}/${BAMFILE}
```

The first issue is that, during step 3, the following error appears in the Clair3 log file:

```[INFO] 3/7 Phase VCF file using Whatshap
This is WhatsHap 1.7 running under Python 3.9.0
[E::vcf_parse_format] Invalid character 'c' in 'AF' FORMAT field at chr4:161631731
ERROR: whatshap error: unable to parse next record
This is WhatsHap 1.7 running under Python 3.9.0
Working on 1 sample from 1 family
```

When I inspect that variant in `pileup.vcf.gz`, using `zcat pileup.vcf.gz | grep -P "chr4\t161631731"` this is what is returned:

```
chr4	161631731	.	C	T	18.26	PASS	P	GT:GQ:DP:AD:AF	0/1:18:19:8,4:0.2105chr4	162256561	.	T	.	18.07	RefCall	P	GT:GQ:DP:AD:AF	0/0:18:24:21:0.8750
```

Essentially, it appears that a newline character is missing, but this causes Clair3 to fail, with no VCF written past `pileup.vcf.gz`.

The second issue, is that even when downsampled BAMs run through Clair3, there are suspiciously few variants called on chr1. For reference, these are the number of variants on each chromosome in `phased_merge_output.vcf.gz` that was run on the 91X (not downsampled BAM):
```
zcat phased_merge_output.vcf.gz | grep -ve "^#" | cut -f1 | sort | uniq -c

 421337 chr1
 263451 chr10
 252921 chr11
 244749 chr12
 202196 chr13
 158896 chr14
 154596 chr15
 158762 chr16
 157615 chr17
 160051 chr18
 114335 chr19
 384804 chr2
 121525 chr20
  78178 chr21
  82399 chr22
 338028 chr3
 353767 chr4
 308948 chr5
 308730 chr6
 294777 chr7
 262141 chr8
 226676 chr9
 138383 chrX
  16041 chrY
```

and for the BAM that was downsampled to 70X coverage:

```
zcat phased_merge_output.vcf.gz | grep -ve "^#" | cut -f1 | sort | uniq -c

  33549 chr1
 361352 chr10
 336461 chr11
 332616 chr12
 246947 chr13
 215390 chr14
 212440 chr15
 213311 chr16
 212008 chr17
 206445 chr18
 159123 chr19
 174195 chr2
 185913 chr20
 110666 chr21
 109394 chr22
 295825 chr3
 338105 chr4
 398279 chr5
 389589 chr6
 382903 chr7
 347337 chr8
 317823 chr9
 224857 chrX
  37439 chrY
```

I'm not entirely sure if these errors are connected, but I've been using Clair3 a lot lately, and these issues only ever been a problem when variant calling on these downsampled BAMs.

**Comment by aquaskyline** [comment]
Are you able to give the latest version a try? I am not sure about what's the exact problem caused by a subsampled BAM, but the 'missing newline' problem seems to be fixed in the latest version.

**Comment by fgfrost** [comment]
I was able to install and run Clair3 v1.0.7 via a conda environment, but still ran into the missing newline problem. This was the error in the log file:

**Comment by aquaskyline** [comment]
Could you please send me the whole log?

**Comment by zhengzhenxian** [comment]
@fgfrost 
There are some errors to open a new subprocess, and a similar issue is found [here](https://github.com/HKU-BAL/Clair3/issues/200). Could you try to reduce the threads and see if the issue persists?
```
Traceback (most recent call last):
  File "/gpfs/gsfs11/users/Udpbinfo/Uploads/Cohorts/Long_read_sequencing/Clair3/scripts/../clair3.py", line 105, in <module>
    main()
  File "/gpfs/gsfs11/users/Udpbinfo/Uploads/Cohorts/Long_read_sequencing/Clair3/scripts/../clair3.py", line 99, in main
    submodule.main()
  File "/gpfs/gsfs11/users/Udpbinfo/Uploads/Cohorts/Long_read_sequencing/Clair3/clair3/CallVariantsFromCffi.py", line 356, in main
    Run(args)
  File "/gpfs/gsfs11/users/Udpbinfo/Uploads/Cohorts/Long_read_sequencing/Clair3/clair3/CallVariantsFromCffi.py", line 63, in Run
    call_variants_from_cffi(args=args, output_config=output_config, output_utilities=output_utilities)
  File "/gpfs/gsfs11/users/Udpbinfo/Uploads/Cohorts/Long_read_sequencing/Clair3/clair3/CallVariantsFromCffi.py", line 119, in call_variants_from_cffi
    tensor, all_position, all_alt_info = CT(args)
  File "/gpfs/gsfs11/users/Udpbinfo/Uploads/Cohorts/Long_read_sequencing/Clair3/preprocess/CreateTensorFullAlignmentFromCffi.py", line 55, in CreateTensorFullAlignment
    candidate_file_path_process = subprocess_popen(shlex.split("gzip -fdc %s" % (full_aln_regions)))
  File "/gpfs/gsfs11/users/Udpbinfo/Uploads/Cohorts/Long_read_sequencing/Clair3/shared/utils.py", line 125, in subprocess_popen
    return Popen(args, stdin=stdin, stdout=stdout, stderr=stderr, bufsize=bufsize, universal_newlines=True)
  File "/vf/users/fuj6/conda/envs/clair3/lib/python3.9/subprocess.py", line 947, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/vf/users/fuj6/conda/envs/clair3/lib/python3.9/subprocess.py", line 1752, in _execute_child
    self.pid = _posixsubprocess.fork_exec(
BlockingIOError: [Errno 11] Resource temporarily unavailable
```

**Comment by fgfrost** [comment]
@zhengzhenxian reducing the threads seemed to solve the issue!It does seem like there's a bit of a oddity in how the problem was fixed, which may be informative. We initially submitted this script to a slurm system on 56 CPU nodes, were we requested the maximum of 56 CPUs (and told Clair3 to use 56). When we tried the lower thread number (28 CPUs), we only requested 28 from the 56 CPU node, and told Clair3 to use all 28, which worked great.

**Comment by aquaskyline** [comment]
Some linux distributions have a lower limit of maximum user processes. It can be checked with `ulimit -u`. Ubuntu usually has a limit large (~1M) enough but CentOS usually sets 1024 or 4096. Users usually see `resource temporary unavailable` if Clair3 is asking for more processes the OS is allowed to provide. Even if one sets using only 1 core, due to the design of Tensorflow, it opens _k_ processes at the start of a tensorflow program, where _k_ is the number of logical cores available. The is one major cause of Clair3's runtime failure saying `ulimit -u` at random points. Say if you set --threads to 56, which means that Clair3 will use 14 concurrent tensorflows (56/4, we set 4 threads per tensorflow by default), the worst case scenario is that all tensorflow program starts at the same time, requiring 56*14=784 users processes, considering a few hundreds of other processes were already used by the OS itself, an 1024 `max users processes` will fail Clair3.

The solution is to set a higher value to `ulimit -u` but that requires admin privilege. If `ulimit -u` cannot be lifted, the only solution is to use a lower `--thread`. This tensorflow trouble is also one of the reasons that we started using PyTorch in later Clair series such as ClairS and ClairS-TO.
